File: C++/hashMap.cpp
https://leetcode.com/problems/design-hashmap/
//using array method
class MyHashMap {
public:

    int data[1000001];

    MyHashMap()
    {
        //this constructor is used to initialise an array data[]
        // with size 1000001 and sets all elements to -1
        // the data[] will serve as the hash map
        for(int i=0; i<1000001; i++)
        {
            data[i] = -1;
        }
        
    }
    

    /*
    This function put()
    takes a key and value as argument, stores the value in the hash map at the position of index key. 
    */
    void put(int key, int value) 
    {
        data[key] = value;
        
    }
    
    /*
    This function get()
    takes a key as an argument and returns the value associated with the key in hash map.
    If the value is not found which means that it does not exist(array element is at index key -1), then it returns -1

    */

    int get(int key) 
    {
        if(data[key] == -1)
        {
            return -1;
        }
        return data[key];
    }
    
    /*
    This function remove()
    takes a key as an argument and removes the value associated with the key in has map.The valye is removes from the hash map by setting the array element if index key to -1
    */
    void remove(int key) 
    {
        if(data[key] != -1)
        {
            data[key] = -1;
        }
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */

File: python/hashMap.py
https://leetcode.com/problems/design-hashmap/
"""
In this case, we would like to solve the problem
by creating a hashmap which has the maximum operation of 1000 times

Concern: Hashmap collision
There are many ways to handle for it
e.g.
1. Open Addressing
2. Chaining 
... etc

We will go for the chaining method, using the linked-list
and creates a dummy nodes infront of the first node

"""
class ListNode:
    def __init__(self, key=None, val=None):
        self.key = key
        self.val = val
        self.next = None

class MyHashMap:

    def __init__(self):
        self.map = [ListNode() for _ in range(1000)]

    def hash(self, key):
        return key % len(self.map)
    
    def put(self, key: int, value: int) -> None:
        #this is a dummy node
        current = self.map[self.hash(key)]
        while current.next:
            #which means the key is found
            if current.next.key == key:
                current.next.val = value
                return
            current = current.next
        current.next = ListNode(key, value)

    def get(self, key: int) -> int:
        #to get the val, we no need to start from dummy node
        #start from the first node
        current = self.map[self.hash(key)].next
        while current:
            if current.key == key:
                return current.val
            current = current.next
        return -1

    def remove(self, key: int) -> None:
        #have to start from the dummy node
        #to avoid accidentally remove the first node
        current = self.map[self.hash(key)]
        #check is dummy and first node empty
        while current and current.next:
            if current.next.key == key:
                current.next = current.next.next
                return
            current = current.next

        


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)